### PR TITLE
feat[graphics]: Added graphical features

### PR DESCRIPTION
**Pull request by MacXxs**
Added:
* ComponentLine
* Shadows with VSM
* Compute shaders
* Trails
* Bloom and tone mapping (Bloom intensity modifyable from the root node inspector window)
* Option to Enable/Disable rigidBodies dd from the Frustum window

Changes:
* ParticleSystem particles now have an isDead flag so the system have been changed in order to fix some "issues" with flickering due to static particle systems that simulate a billboard (such as the reticle on top of enemies).

Commands:
* F1 to toggle bloom
* F2 to change tone-mapping
* F3 to toggle shadows on/off
* F4 to toggle VSM on/off
* F5 to change the rendered buffer texture

### CODE REVIEW

**Comment by UriMurcia** [comment]
I was testing different values for bloom intensity but it looks quite the same. Would be nice to be able to get 0 intensity so we can calibrate it better
image
image

**Comment by UriMurcia** [comment]
Some bugs happened while testing trails:
image
image
image

What happened in the second image maybe can be avoided but I thought would be interesting showing it.

**Comment by UriMurcia** [comment]
Also would be great to be nice to be able to have lower numbers than 1 for width and Ratio width
image

**Comment by MacXxs** [comment]
> Some bugs happened while testing trails: image image image
> 
> What happened in the second image maybe can be avoided but I thought would be interesting showing it.

Can you explain how did you break it or upload a video of the bug?

File: Game/Scripts/HealthSystem.cpp [hunk 1]
```diff
@@ -9,16 +9,22 @@
...
 HealthSystem::HealthSystem() : Script(), currentHealth(100), maxHealth(100), componentAnimation(nullptr), 
-	isImmortal(false), enemyParticleSystem(nullptr), attackScript(nullptr)
+isImmortal(false), enemyParticleSystem(nullptr), attackScript(nullptr), hitEffectDuration(0.f), hasTakenDamage(false)
 {
 	REGISTER_FIELD(currentHealth, float);
 	REGISTER_FIELD(maxHealth, float);
 	REGISTER_FIELD(isImmortal, bool);
 	REGISTER_FIELD(enemyParticleSystem, GameObject*);
+
+	REGISTER_FIELD(meshEffect, MeshEffect*);
```
[the remark this answers is not in the document]
Comment (PaablooCH) [reply Game/Scripts/HealthSystem.cpp:27]: I followed the suggestions in the original pr and each script that needs the effectMesh needs a different one and they are stored in a gameobject outside the main gameobject (like strongAttack, i don't remember the name) so doing what you suggest couldn't be an option.

File: Game/Scripts/HealthSystem.cpp [hunk 2]
```diff
@@ -44,6 +50,10 @@ void HealthSystem::Start()
 		maxHealth = currentHealth;
 	}
 
+	meshEffect->FillMeshes(owner);
+	meshEffect->ReserveSpace(1);
+	meshEffect->AddColor(float3(1.f, 0.f, 0.f));
```
[the remark this answers is not in the document]
Comment (PaablooCH) [reply Game/Scripts/HealthSystem.cpp:55]: i don't know if scripts support vector of float3, if you can give an example i will implemented

> Also we could look at the number of colors at the start and not be forced to do it from code

what do You mean with this?

**Comment by PaablooCH** [comment]
> I was testing different values for bloom intensity but it looks quite the same. Would be nice to be able to get 0 intensity so we can calibrate it better image image

this will be different once we merge #673, it will have a global intensity intead of a particullary for each material. Regarding the 0 intensity, my opinion is that if you want that value it would be preferable to delete the emissive texture from that object, but as i said we will have a global intensity.

**Comment by UriMurcia** [comment]
> > I was testing different values for bloom intensity but it looks quite the same. Would be nice to be able to get 0 intensity so we can calibrate it better image image
> 
> this will be different once we merge #673, it will have a global intensity intead of a particullary for each material. Regarding the 0 intensity, my opinion is that if you want that value it would be preferable to delete the emissive texture from that object, but as i said we will have a global intensity.

Okay, that's good, but I think it is also good to have a value of intensity in each material, like in Unity. And yes, 0 intensity is useless, but from 0 to 1 is useful

**Comment by UriMurcia** [comment]
> > Some bugs happened while testing trails: image image image
> > What happened in the second image maybe can be avoided but I thought would be interesting showing it.
> 
> Can you explain how did you break it or upload a video of the bug?

What I did was playing with high and max values for the variables in the trail component.

**Comment by PaablooCH** [comment]
> Also would be great to be nice to be able to have lower numbers than 1 for width

done. 

>  and Ratio width

Ratio widht is a percentatge can not be more than 1.

> . And does the Play button do something different from Enabled? Can be deleted from the inspector?

No it can't, it's used to test the trail without script

**Comment by PaablooCH** [comment]
> > > Some bugs happened while testing trails: 
> > > What happened in the second image maybe can be avoided but I thought would be interesting showing it.
> > 
> > 
> > Can you explain how did you break it or upload a video of the bug?
> 
> What I did was playing with high and max values for the variables in the trail component.

probably the 3 are related because for some reason after the asynchronous merge, the buffers are no longer created

**Comment by MacXxs** [comment]
> > > I was testing different values for bloom intensity but it looks quite the same. Would be nice to be able to get 0 intensity so we can calibrate it better image image
> > 
> > 
> > this will be different once we merge #673, it will have a global intensity intead of a particullary for each material. Regarding the 0 intensity, my opinion is that if you want that value it would be preferable to delete the emissive texture from that object, but as i said we will have a global intensity.
> 
> Okay, that's good, but I think it is also good to have a value of intensity in each material, like in Unity. And yes, 0 intensity is useless, but from 0 to 1 is useful

In Unity the bloom intensity is applied from a postprocess layer that applies bloom to every object with emissive, and applies the same bloom for each one of them. The only thing that you can modify in unity it's the intensity of the emissive textures perse, no its bloom. 

e2361a0ec9c1bd1e031e141a522be33f

**Comment by gerard-queralt** [comment]
> Commands:
> 
> * F2 to change tone-mapping
> * F3 to toggle shadows on/off
> * F4 to toggle VSM on/off
> * F5 to change the rendered buffer texture

If you don't mind, could you add this to the `Controls` section of the readme?

**Review by PaablooCH — DISMISSED** [review]

**Review by PaablooCH — APPROVED** [review]

**Review by AlexisDieu — APPROVED** [review]

**Pull request merged**